### PR TITLE
Ensure `Matrix` * `(Sym)Tridiagonal` is dense

### PR DESCRIPTION
**Pull request by mcognetta**
This PR ensures that `Matrix` * `(Sym)Tridiagonal` outputs a dense matrix. Before, it was outputting a `SparseMatrixCSC`, despite the resulting matrix being dense.

This PR currently only implements this optimization for the `Matrix` type, so as to avoid method ambiguities, but it is possible that there are other type combinations that suffer from the same problem.

------------------------------------------------------------------------------------
A basic benchmark, showing a ~150x speedup:

Before:

```julia
julia> t = Tridiagonal(rand(1000, 1000)); s = SymTridiagonal(rand(1000), rand(999)); y = rand(1000, 1000);

julia> @btime $y * $t;
  256.924 ms (41 allocations: 18.01 MiB)

julia> @btime $y * $s;
  260.044 ms (41 allocations: 18.01 MiB)
```

This PR:

```julia
julia> t = Tridiagonal(rand(1000, 1000)); s = SymTridiagonal(rand(1000), rand(999)); y = rand(1000, 1000);

julia> @btime $y * $t;
  1.477 ms (2 allocations: 7.63 MiB)

julia> @btime $y * $s;
  1.469 ms (2 allocations: 7.63 MiB)
```

### CODE REVIEW

**Comment by mcognetta** [comment]
`buildbot/tester_freebsd64` check failure seems unrelated.

**Comment by dkarrasch** [comment]
Could it be that this is already handled by #43127?

**Comment by dkarrasch** [comment]
> Could it be that this is already handled by #43127?

Yes, it is, I checked. Good to know that there are also performance benefits from splitting LinearAlgebra and SparseArrays. Maybe we keep this PR for the added tests until #43127 is merged?

**Comment by dkarrasch** [comment]
BTW, the same change should be made for `Bidiagonal`, and also that is already "included" in #43127.

**Comment by mcognetta** [comment]
Ah, thanks for the info. I didn't realize this was in the scope of that PR. I'll keep it around until #43127 is merged.

File: stdlib/LinearAlgebra/src/special.jl [hunk 1]
```diff
@@ -294,6 +294,12 @@ rmul!(A::AbstractTriangular, adjB::Adjoint{<:Any,<:Union{QRCompactWYQ,QRPackedQ}
 *(adjA::Adjoint{<:Any,<:Union{QRCompactWYQ, QRPackedQ}}, B::BiTriSym) =
     lmul!(adjA, copyto!(Array{promote_type(eltype(adjA), eltype(B))}(undef, size(B)...), B))
 
+function (*)(A::Matrix, B::Union{SymTridiagonal, Tridiagonal})
```
Comment (stevengj) [inline stdlib/LinearAlgebra/src/special.jl:297]:
```suggestion
function (*)(A::StridedMatrix, B::Union{SymTridiagonal, Tridiagonal})
```
?

**Comment by stevengj** [comment]
Is there a similar method for `tridiagonal * matrix` that also needs to be fixed?

**Comment by mcognetta** [comment]
@stevengj Not for Tridiagonal * Dense, but perhaps for others. I think we are getting pretty close to catching all the defined structured matrix multiplication optimizations (though it always seems like one slips through the cracks).

**Comment by dkarrasch** [comment]
This requires a rebase, then this should work and we can merge.

**Comment by mcognetta** [comment]
Ok, that did not work.

I am going to close this PR and open a new one with just the tests.

**Comment by mcognetta** [comment]
Opened #43549 to resolve this.

**Pull request closed without merging**